### PR TITLE
CORE-15757  flow session refactor

### PR DESCRIPTION
**Pull request by LWogan**
Flow Session Refactor to reduce the total messages exchanged between session counterparties. 

SessionInit info can now be piggybacked onto SessionData messages.

SessionAcks are removed.

SessionContextProperties are piggybacked onto any SessionEvents as SessionConfirm is not always sent.

Add new initiateFlow API to allow the initiator to define whether a SessionClose message should be sent by the initiated party when their flow finishes or when they call close. Whichever comes first.

Runtime-OS:
https://github.com/corda/corda-runtime-os/pull/4446
E2E Tests
https://github.com/corda/corda-e2e-tests/pull/218

### CODE REVIEW

File: application/src/main/java/net/corda/v5/application/messaging/FlowMessaging.java [hunk 1]
```diff
@@ -122,7 +147,7 @@ public interface FlowMessaging {
      */
     @Suspendable
     @NotNull
-    FlowSession initiateFlow(@NotNull MemberX500Name x500Name, @NotNull FlowContextPropertiesBuilder flowContextPropertiesBuilder);
+    FlowSession initiateFlow(@NotNull MemberX500Name x500Name, @NotNull Boolean requireClose, @NotNull FlowContextPropertiesBuilder flowContextPropertiesBuilder);
```
Comment (chrisr3) [inline application/src/main/java/net/corda/v5/application/messaging/FlowMessaging.java]: By `@NotNull Boolean` I assume you really mean `boolean`?

File: application/src/main/java/net/corda/v5/application/messaging/FlowMessaging.java [hunk 2]
```diff
@@ -84,6 +84,28 @@ public interface FlowMessaging {
...
+     * @param x500Name The X500 name of the member to communicate with.
+     * @param requireClose When set to true, the initiated party will send a close message after calling FlowSession.close()
+     *                     and the initiating party will suspend and wait to receive the message when they call FlowSession.close().
+     *                     When set to false the session is marked as terminated immediately when close() is called.
+     *
+     * @return The session.
+     */
+    @Suspendable
+    @NotNull
+    FlowSession initiateFlow(@NotNull MemberX500Name x500Name, Boolean requireClose);
```
Comment (chrisr3) [inline application/src/main/java/net/corda/v5/application/messaging/FlowMessaging.java]: Also `boolean` rather than `Boolean`?

**Comment by LWogan** [comment]
@chrisr3 as mentioned in the draft text this PR is not for review yet

**Comment by chrisr3** [comment]
> @chrisr3 as mentioned in the draft text this PR is not for review yet

This change looks like [something unexpected that is difficult just to ignore](https://www.youtube.com/watch?v=d4HjDc44_VQ)...

**Comment by LWogan** [comment]
> > @chrisr3 as mentioned in the draft text this PR is not for review yet
> 
> This change looks like [something unexpected that is difficult just to ignore](https://www.youtube.com/watch?v=d4HjDc44_VQ)...

I may have made some updates this morning.... completely unrelated to your comments.... 👀

File: application/src/test/java/net/corda/v5/application/messaging/FlowMessagingJavaApiTest.java [hunk 3]
```diff
@@ -4,8 +4,7 @@
 import org.assertj.core.api.Assertions;
 import org.junit.jupiter.api.Test;
 
-import static org.mockito.ArgumentMatchers.any;
-import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.ArgumentMatchers.*;
```
Comment (ben-millar) [inline application/src/test/java/net/corda/v5/application/messaging/FlowMessagingJavaApiTest.java]: Minor: wildcard import

**Review by JamesHR3 — DISMISSED** [review]

**Review by JamesHR3 — DISMISSED** [review]

**Review by JamesHR3 — APPROVED** [review]

**Pull request merged**